### PR TITLE
Add statistics counter for generating a time histogram

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
A timing statistics counter for any sort of transaction that takes a
microsecond to a second to complete. Collects the following statistical
information:

- a logarithmic time histogram with bins from ≥1µs to <1s, three bins per
  power of ten in a  stepping of 1 .. 2 .. 5 .., plus one bin for each <1µs
  and ≥1s,

- the total number of transactions and the aggregated total completion time.

### CODE REVIEW

File: src/ocean/util/TimeHistogram.d [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    To reset all counters to zero use `TimeHistogram.init`.
+
+    copyright: Copyright (c) 2018 dunnhumby Germany GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module ocean.util.TimeHistogram;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/TimeHistogram.d]: Should probably live in https://github.com/sociomantic-tsunami/ocean/tree/v4.x.x/src/ocean/math, along with `BinaryHistogram`?

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/TimeHistogram.d]: Moved.

File: src/ocean/math/TimeHistogram.d [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+
+        Params:
+            us = the transaction completion time in microseconds
+
+        Returns:
+            us
+
+    ***************************************************************************/
+
+    ulong countMicros ( ulong us )
```
Comment (gavin-norman-sociomantic) [inline src/ocean/math/TimeHistogram.d]: `public`.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Apart from one tiny comment, LGTM.

**Comment by nemanja-boric-sociomantic** [comment]
I've added `public` also to `total_time_micros`, `count` and `bins`. AFAIR these are also supposed to be used from the outside.

**Comment by gavin-norman-sociomantic** [comment]
Yeah, this struct just trusts its user to not mess with its internals. I think that's fine.

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Pull request merged**